Update: Proserv Style Guide Updates (#63)
dbt-labs/corp: dbt_style_guide.md

# dbt Style Guide

## Model Naming

- All objects should be plural.  
  Example: `stg_stripe__invoices.sql` vs. `stg_stripe__invoice.sql`

> - All objects should have a prefix to indicate their purpose in the flow:

Author: I instead moved it to the "dbt Conventions" section - [found in this commit](https://github.com/dbt-labs/corp/pull/63/commits/4738259920c600b9f5c10953a0c59ae42e70a77a)

  - `base_` is uncommon, but indicates a process needed before `stg_` - typically when multiple sources are rarely used independently. The `stg_` model is then used to join or union the sources together before more robust transformations happen.  
     
     Example: Creating one table for all cleaned location data.
     - `base_location__addresses.sql`, `base_location__countries.sql`, and `base_location__states.sql`  would serve to clean the data and be the 1:1 relationship between the source.
     - `stg_location__locations.sql` would serve to join all location-related data as one entity.

>   - `stg_` is used to clean and standardize data before being fundamentally changed in downstream modeling.

Reviewer: Maybe add "in the absence of a base_ layer, stg_ models have a 1:1 relationship with the source" or something similar

Author: @graciegoheen there's been a change to where the definitions are now located under `dbt conventions`. Give that a read and tell me how you feel about it!

>   - `int_` is used to indicate a step towards creating a final model surfaced to stakeholders.

Reviewer: Maybe "captures intermediate transformations to be used by multiple downstream fct_ or dim_ models" or something similar

Author: Will think about how to reword this on Friday! My only qualm is using the word "intermediate" - I feel like it's the same thing as using the same term you're trying to define in a definition!

Author: @graciegoheen Can you take a look at [this](https://github.com/dbt-labs/corp/pull/63/commits/02e749ad0b2f9ed8848cf7d9978a9785befb073f) change and tell me if it works for you?

>   - `fct_` is used to indicate a final data set surfaced to stakeholders, and flags data which is in the form of an immutable event stream.

Author: Missed one! [Here it is](https://github.com/dbt-labs/corp/pull/63/commits/5f4b1e3a58e9292be52de042578ba7ceaa2c8cf1)

  - `dim_` is used to indicate a final data set surfaced to stakeholders, and flags data which is used to describe an entity.  
  - `seed_` is used to indicate that the data set was built from a CSV file
  using `dbt seed`.

> - All models should use the naming convention `<type/dag_stage>_<source/topic>__<additional_context>`. See [this article](https://docs.getdbt.com/blog/stakeholder-friendly-model-names) for more information.

Reviewer: Here you call it the "dag_stage" but in the list above it's called "purpose in the flow" - can we connect the two?

Author: Done! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/1c973622c93fc2736e3e7b9fbbb2c06ba708a915)

  - Within the **marts** and **intermediate** folders, `__<additional_context>` is optional. 
  - Models in the **staging** folder should use the source's name as the `<source/topic>` and the entity name as the `additional_context`.

    Examples:
    - seed_snowflake_spend.csv
    - base_stripe__invoices.sql
    - stg_stripe__customers.sql
    - int_payments.sql
    - int_customers__unioned.sql
    - fct_orders.sql

## Model Organization  

> Our models (typically) fit into three main categories: staging, marts, and base/intermediate.  

Reviewer 2: I think that `intermediate` models are in all (or most of) the dbt projects I have seen. `base` are much rarer. So, we could rewrite saying that it is often 3 (removing `base`) but could be 4 with `base`.

> Our models (typically) fit into three main categories: staging, marts, and base/intermediate.  

Reviewer: Can we connect this back to the prefixes?

> Our models (typically) fit into three main categories: staging, marts, and base/intermediate.  

Author: Taking in to consideration everything said, I feel like [this change](https://github.com/dbt-labs/corp/pull/63/commits/1c973622c93fc2736e3e7b9fbbb2c06ba708a915) is really what we need. Happy to change this back and just redo the wording if no one agrees! @graciegoheen can you also check that this covers your ask?

> For more detail about why we use this structure, check out [this discourse post](https://discourse.getdbt.com/t/how-we-structure-our-dbt-projects/355). The file and naming structures are as follows:

Author: I'm going to take out the example file structure and just link to that page - The one thing to consider there is that when adding a style guide to a client's project, that page is liable to change so they may need to do some major work and write out their own preferred structure. However, in our best interests I think one place to maintain makes more sense for less overhead. 

Also going to shout out that I don't fully agree with the structure in the article (for example, I don't like _finance__models.yml and prefer _models_finance.yml so that the important info is first) but I'm not really saying this for any reason but to have someone listen 🥲 Tomato Tuhmahto

> For more detail about why we use this structure, check out [this discourse post](https://discourse.getdbt.com/t/how-we-structure-our-dbt-projects/355). The file and naming structures are as follows:

Author: [Updated](https://github.com/dbt-labs/corp/pull/63/commits/17817a412bb10d9b177e6bcd28ce65182f3ce6ff)

```
├── dbt_project.yml
└── models
    ├── marts
    |   ├── core
    |   |   ├── _docs_core.md
    |   |   ├── _models_core.yml
    |   |   ├── dim_customers.sql
    |   |   └── fct_orders.sql
    |   └── intermediate
    |       ├── _models_int_core.yml
    |       ├── int_payments.sql
    |       ├── int_customers__unioned.sql
    |       └── int_customers__grouped.sql
    └── staging
        └── stripe
            ├── base
            |   ├── _models_base_stripe.yml
            |   └── base_stripe__invoices.sql
            ├── _docs_stripe.md
            ├── _models_stripe.yml
            ├── _sources_stripe.yml
            ├── stg_stripe__customers.sql
            └── stg_stripe__invoices.sql
```
## Model configuration

> - Model configurations at the [group level](https://docs.getdbt.com/reference/model-configs#configuring-directories-of-models-in-dbt_projectyml) should be considered (and if applicable, applied) first.

Reviewer: Maybe "folder level" instead of "group level"

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/94a41d4acbebbb12a8c7455b87a5abf3624aaa4f)

- More specific configurations should be applied at the model level [using one of these methods](https://docs.getdbt.com/reference/model-configs#apply-configurations-to-one-model-only).
- Models within the `marts` folder should be materialized as `table` or `incremental`.
  - By default, `marts` should be materialized as `table` within `dbt_project.yml`.
  - If switching to `incremental`, this should be specified in the model's configuration.

## dbt conventions
- Only `base_` and `stg_` models should select from [sources](https://docs.getdbt.com/docs/building-a-dbt-project/using-sources)
- Models not within the `staging` folder should select from [refs](https://docs.getdbt.com/reference/dbt-jinja-functions/ref).

## Testing

> - At a minimum, `unique` and `not_null` tests should be applied to the assumed primary key of each model.

Reviewer: How about "expected primary key" instead?

## Naming and field conventions

- Schema, table and column names should be in `snake_case`.

- Limit use of abbreviations that are related to domain knowledge. An onboarding
  employee will understand `current_order_status` better than `current_os`.

- Use names based on the _business_ terminology, rather than the source terminology.

- Each model should have a primary key that can identify the unique row.

- The primary key of a model should be named `<object>_id`, e.g. `account_id` – this makes it easier to know what `id` is being referenced in downstream joined models.

- If a surrogate key is created, it should be named `<object>_sk`.

> - For `base` or `staging` models, columns should be ordered in categories, where identifiers are first and date/time fields are at the end.  

Reviewer 2: I don't think that all datetime would be metadata though. In my mind, `loaded_at` would be metadata, put `order_created_at` wouldn't.

> - For `base` or `staging` models, columns should be ordered in categories, where identifiers are first and date/time fields are at the end.  

Author: Done [commit](https://github.com/dbt-labs/corp/pull/63/commits/c1ca3acc174ca283077820d3106479fbc5ec8832)

  Example:
  ```sql
  transformed as (

      select

        -- ids
        order_id,
        customer_id,

        -- dimensions
        order_status,
        is_shipped,

        -- measures
        order_total,

>         -- metdata

Author: Done [commit](https://github.com/dbt-labs/corp/pull/63/commits/c1ca3acc174ca283077820d3106479fbc5ec8832)

        created_at,
        updated_at,
        _sdc_batched_at

      from source

  )
  ```

- Date/time columns should be named according to these conventions:
  - Timestamps: `<event>_at`
    Format: UTC  
    Example: `created_at`
  
  - Dates: `<event>_date`  
    Format: Date  
    Example: `created_date`

> - Booleans should be prefixed with `is_` or `has_`.

Reviewer 3: Agreed. Adding a suggestion block to make committing easier.
```suggestion
- Booleans should be prefixed with `is_` or `has_`.
Example: `is_active_customer` and `has_admin_access`

```

> - Booleans should be prefixed with `is_` or `has_`.

Author: Done [commit](https://github.com/dbt-labs/corp/pull/63/commits/4120de8f0221b6e4dd76c3f3120dc50feef1d90c)

- Price/revenue fields should be in decimal currency (e.g. `19.99` for $19.99; many app databases store prices as integers in cents). If non-decimal currency is used, indicate this with suffix, e.g. `price_in_cents`.

- Avoid using reserved words (such as [these](https://docs.snowflake.com/en/sql-reference/reserved-keywords.html) for Snowflake) as column names.

- Consistency is key! Use the same field names across models where possible.  
Example: a key to the `customers` table should be named `customer_id` rather than `user_id`.

## CTEs

For more information about why we use so many CTEs, check out [this glossary entry](https://docs.getdbt.com/terms/cte).

- Where performance permits, CTEs should perform a single, logical unit of work.

- CTE names should be as verbose as needed to convey what they do.

- CTEs with confusing or noteable logic should be commented with SQL comments, as you would with any complex functions.

- CTEs that are duplicated across models should be pulled out and created as their own models.

> - CTEs and SQL within a model should follow this structure:

Reviewer 4: I think this section should be moved below the next two bullet points. We don't define what an import CTE is before showing the order of it.

Reviewer 5: +1

Reviewer 6: +1

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/a75fe2eb326e6bcc87479b1256b6b6bd84533446)

  - `with` statement
  - Import CTEs
  - Logical CTEs
  - Final CTE
  - Simple `select` statement

- All `{{ ref('...') }}` statements should be placed in CTEs at the top of the file ("import" CTEs)

- SQL should end with a simple select statement. All other logic should be contained within CTEs.  

>   Example: `select * from final`

Author: Done, but not sure which commit this was in! Just look at the updated version.

> - Where applicable, opt for filtering within import CTEs over filtering within logical CTEs.

Reviewer: Do we define what a logical cte is? I see we name `"import" CTEs` - can we do the same for final and logical?

Author: @graciegoheen Do you have a good example of what you're thinking here?

Reviewer 7: I believe this just means if you are joining some tables and always filtering something that could be moved up to the top, like:
```
with
customers as (
  select * from {{ ref('stg_customers') }} where is_deleted is false
)
```

Author: @graciegoheen Can you see if [this new structure](https://github.com/dbt-labs/corp/pull/63/commits/ad31e8a610c7ffb17cbf2bff901c7794f5c725d5) helps?

- ### Example CTE

  ``` sql
  with

  events as (

      ...
      where not is_deleted

  ),

  -- CTE comments go here
  events_joined as (

      ...

  )

  select * from events_joined
  ```

## SQL style guide
- **DO NOT OPTIMIZE FOR A SMALLER NUMBER OF LINES OF CODE.**  
  New lines are cheap, brain time is expensive; new lines should be used within  
  reason to produce code that is easily read.

- Use trailing commas

- Indents should be four spaces 

> - When dealing with long `when` or `where` clauses, predicates should be on a new

Reviewer: Is this not already covered by the section below:

> - When filtering by multiple clauses, each clause and expression should be on a new line.  

Might be good to connect these two?

Author: Removed the last one! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/d743e54439bf4b269d94927a6c48bb937e9eb338)

  line and indented.

- Lines of SQL should be no longer than 80 characters and new lines should be used to ensure this.  
  Example:
  ```sql
  sum(
    case when order_status = 'complete' then order_total end
  ) as monthly_total,

  {{ get_windowed_values(
        strategy='sum',
        partition='order_id',
        order_by='created_at',
        column_list=[
          'final_cost'
        ]
  ) }} as total_final_cost
  ```

> - Field names and function names should all be lowercase

Reviewer: file names too? Not sure where that would go - but see clients all the time use capitals in their file names which then makes it hard to use selector syntax

Author: Really good point! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/aa08fe9b6842b24aafb09db26db82a76e9f0cbb1)

- The `as` keyword should be used when aliasing a field or table

- Fields should be stated before aggregates / window functions

- Aggregations should be executed as early as possible before joining to another table.

> - Ordering and grouping by a number (eg. group by 1, 2) is preferred over listing the column names (see [this rant](https://blog.getdbt.com/write-better-sql-a-defense-of-group-by-1/) for why). Note that if you are grouping by more than a few columns, it may be worth revisiting your model design.

Reviewer 5: Can we maybe add a TL;DR for the article here?

Reviewer 6: also a good place to mention dbt_utils.group_by function

Author: @SamHarting not adding a TL;DR because many reasons are listed in the article in which I think a TL;DR would be more than a one-liner. We also didn't do that (except for stating the main points) for any of our other links! If you have something you had in mind we can definitely start up a convo about it/change it!

@lbenezra-FA Added the suggestion for the group_by function. [Commit](https://github.com/dbt-labs/corp/pull/63/commits/c18398f700c0bcc8e175863fe5b044912c784a1b)

- Prefer `union all` to `union` [*](http://docs.aws.amazon.com/redshift/latest/dg/c_example_unionall_query.html)

- Avoid table aliases in join conditions (especially initialisms) – it's harder to understand what the table called "c" is compared to "customers".

- If joining two or more tables, _always_ prefix your column names with the table alias. If only selecting from one table, prefixes are not needed.

> - Be explicit about your join (i.e. write `inner join` instead of `join`). `left joins` are normally the most useful, `right joins` often indicate that you should change which table you select `from` and which one you `join` to.

Reviewer: ```suggestion
- Be explicit about your join (i.e. write `inner join` instead of `join`). `left joins` are normally the most common, `right joins` often indicate that you should change which table you select `from` and which one you `join` to.

```

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/4b7f51a09054da94c9050527023b01d42f3e2e9e)

- ### Example SQL
  ```sql
  with

  my_data as (
      select * from {{ ref('my_data') }}
      where not is_deleted
  ),

  some_cte as (
      select * from {{ ref('some_cte') }}
  ),

  some_cte_agg as (
      select
          id,
          sum(field_4) as total_field_4,
          max(field_5) as max_field_5
      from some_cte
      group by 1
  ),

  final as (
      select [distinct]
          my_data.field_1,
          my_data.field_2,
          my_data.field_3,

          -- use line breaks to visually separate calculations into blocks
          case
              when my_data.cancellation_date is null
                and my_data.expiration_date is not null
                then expiration_data
              when my_data.cancellation_date is null
                then my_data.start_date + 7
              else my_data.cancellation_date
          end as cancellation_date,

          some_cte_agg.total_field_4,
          some_cte_agg.max_field_5
      from my_data
      left join some_cte_agg  
          on my_data.id = some_cte_agg.id
      where 
        my_data.field_1 = 'abc'
        and (
            my_data.field_2 = 'def'
            or my_data.field_2 = 'ghi'
        )

>       having count(*) > 1

Reviewer 4: this is syntactically incorrect, you need a `group by` to use a `having`

Reviewer 5: +1

Author: Good catch! How's [this instead](https://github.com/dbt-labs/corp/pull/63/commits/c25dbfe85ad4c5e5b31262683076345d043412d3)?

  )

  select * from final
  ```

> ## YAML style guide

Reviewer 2: Could we add one bullet point:
"- Sources and models in a single `.yml` file should be sorted alphabetically to make it easier to find information in larger files"

I was doing that in past projects and it was great to find where to look in a big YAML file

Reviewer 7: +1

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/01bdeb2a5d3fe7e0e737d78a22628ef7dee6eb4a)

- Every subdirectory contains their own `.yml` file(s) which contain configurations for the models within the subdirectory.

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Reviewer 6: +1

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Author: [Resolved](https://github.com/dbt-labs/corp/pull/63/commits/7420d40c8f6e05be865bebe8f8de6431fe3680e6)

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Author: The duplicate entry was removed, and I think that we should refrain from explaining _another_ practice that isn't the default (so that would change to "We recommend creating ...". The example yml location can already be found in the Best Practices article we linked above, so I personally feel that we should omit a file structure example. 

Gonna leave this one open to see what others think! The guide is getting pretty long so another thing I have on the brain is trying to avoid being too verbose, but since I'm in this headspace of many changes I'd really prefer if everyone else came at this from a different angle!

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Reviewer 7: My reasoning for the example is because I had to read the sentence a few times to figure out what it meant. If someone is less familiar than us, they might need the example. 

I'd say this is a great time to be verbose!

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Reviewer 5: Vote to keep in the examples! I always appreciate them and will just scroll past if I don't need the extra help.

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Reviewer 4: I'm good either way on an example or not, but if we keep the example I think it needs tweaked a bit. Right now it's showing the `jaffle_shop` folder as an empty folder, with the files at the same level as the folder. The files from `_jaffle_shop__models.yml` to `stg_jaffle_shop__payments.sql` should be one level under `jaffle_shop`.

> - A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

Author: Added yml example and fixed the filetree [here](https://github.com/dbt-labs/corp/pull/63/commits/26274f7b5939791e82afcbd4af10498086a03ed8)!

- YAML files should be prefixed with an underscore to keep it at the top of the subdirectory.  

>   Example: `__sources_core.yml`, `_models_core.yml`

Reviewer 7: ```suggestion
  Example: `_core__sources.yml`, `_core__models.yml`

```

Reviewer 3: Alternate suggestion to be in alignment with our [examples above](https://github.com/dbt-labs/corp/pull/63/files#diff-79ee8d7d102b23eb33883f979da5a6db69e0e3336e4397a5ff2174382508f342R43):
```suggestion
  Example: `_sources_core.yml`, `_models_core.yml`

```

Author: I think that the _<config>_<location>.yml was the convention we'd voted on previously. Will change it to that but everyone is welcome to chime in here!

Author: Give me your 🌶️ takes. [Commit](https://github.com/dbt-labs/corp/pull/63/commits/bb03059b4295b374d2f0bdd598960696b009b324)

> - YAML files should be named for the top-level configuration it contains.  
>   Examples: `_sources_core.yml`, `_models_core.yml`

Reviewer 7: ```suggestion
- YAML files should be named for the top-level configuration it contains.  
  Examples: `_jaffle_shop__models.yml`, `_jaffle_shop__sources.yml`

```

Reviewer: ^That is not consistent with the prior examples.

Reviewer 7: Generally, names should go `general_to_specific` so that they alphabetize correctly. 
You want a list like:
```
core_models
core_sources
customers
customers__pivoted
orders
orders__pivoted
```
not like:
```
customers
models_core
orders
pivoted__customers
pivoted__orders
sources_core
```

Reviewer 7: similar to date formats
good:
```
2019-11-23
2020-04-12
2022-08-18
```
bad
```
04-12-2020
08-18-2022
11-23-2019
```

Author: Again I think we voted on having `_<config>_<thing>.yml` early on (possibly in Slack? Or in previous comments here?) Maybe we should all vote on this again. For config files I personally prefer the specific > general for this use case. For example, I wouldn't want `customers__jaffle_shop_stg.sql` as my model file name.

Author: Although, I guess it depends on how you look at it and how you're structuring it. If I put all my yml configs / .mds for my models at the staging level but had subfolders as we typically do, then the more general would be the _docs, _sources, or _models and the specific would be the folder:
```
_docs_asana.md
_docs_jaffle_shop.md
_docs_snowplow.md
_source_asana.yml
_source_jaffle_shop.yml
_source_snowplow.yml
```

Reviewer 7: Are you disagreeing with yourself with this example?
> For example, I wouldn't want customers__jaffle_shop_stg.sql as my model file name.
customers is specific, it is the name of the table, which is more specific than the schema
jaffle_shop is the less specific, it is the bucket which holds customers
stg_ is general ie. is it staging or marts, and a folder up.

the way we write it is most general to most specific: `stg_jaffle_shop__customers`

Reviewer 7: >then the more general would be the _docs, _sources, or _models and the specific would be the folder:

I think you're thinking about this backwards. 🤔

Reviewer 5: my preference is `_directory_models.yml`, `_directory_sources.yml `, `_directory_docs.md`, `_directory_exposures.yml`. In theory you should only have one of each of these in each subdirectory so there shouldn't be any issues with ordering.

Reviewer 5: ^ A reminder that this vote is only for .yml files and that there can be different conventions for .yml files as opposed to models. Also that the underscore in front is vital to proper file structure.

Reviewer 4: I agree with Sam and Pat here. I think the layout Sam provided in his example above is the correct generic > specific setup. This also aligns with the current guide on ["How we structure our dbt projects"](https://docs.getdbt.com/guides/best-practices/how-we-structure/1-guide-overview#:~:text=_jaffle_shop__models.yml)

Author: Changed the conventions to use what's in the article and additionally changed it a bit to include .md in this section, since most of the rules apply. View it [here](https://github.com/dbt-labs/corp/pull/63/commits/42e7f44c0e857f87286595855fc8b3518e9a8944)!

> - Indents should use four spaces.

Reviewer 4: the example below uses two spaces instead of 4. I think we should update this to be 2 spaces

Reviewer 7: +1 thought we prefer 2 for yml and 4 for SQL (and by extension Jinja)

Author: I think we need another vote on this because I'm terrible at finding conversations. There was a convo / vote on this, and I thought it was 2 all around. But yes, I think yml is definitely 2 - not sure what the vote is on SQL anymore! Guessing that because I changed the code within the examples, that there was a vote for 2 on SQL as well.

Reviewer 5: My vote is however many spaces tab is for both... lazy I guess 🤷🏻‍♂️

Reviewer 2: I don't really care for SQL (2 or 4), but definitely 2 for YML

Author: Fixed [here](https://github.com/dbt-labs/corp/pull/63/commits/b24fa76bae298c05105001532ad70c4b885a0192)!

- List items should be indented.

- Use a new line to separate list items that are dictionaries, where appropriate.

> - Lines of YAML should be no longer than 80 characters.

Reviewer 5: I've always wondered, with how weird yml is about white space, is this something we recommend? I know in our description we can do ': > ' but if it isn't a field with that option can they create that white space without breaking something?

Reviewer 7: I actually just tested this today with `accepted_values` and you can indeed do things like this and it will still work:
```
        tests:
          - accepted_values:
              values: [
                'AFRICA',
                'MIDDLE EAST',
                'ASIA','EUROPE','AMERICA']
```

As for descriptions, `:>` works, as does something like: 
```
version: 2

models:
  - name: dim_customers
    description: "
    This is a multi
    paragraph description of
    Customer dimensions table
    "
    columns:
      - name: customer_id
```

Reviewer 5: Awesome, thanks Pat.

          description: When the event occurred in UTC (eg. 2018-01-01 12:00:00)
          tests:
            - not_null

        - name: user_id
          description: >
            The user id of the visitor to the site.
            This only populates when the user logs in - this user id can be  
            joined to the Jaffle Shop Users data.
          tests:
            - not_null
            - relationships:
                to: ref('users')
                field: id
  ```

## Jinja style guide

- Jinja delimiters should have spaces inside of the delimiter between the brackets and your code.  
  Example: `{{ this }}` instead of `{{this}}`

- Use [whitespace control](https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control) to make compiled SQL more readable.

- An effort should be made for a good balance in readability for both templated 
and compiled code. However, opt for code readability over compiled SQL readability
when needed.

- A macro file should be named after the _main_ macro it contains.

- A file with more than one macro which will be used independently should follow
these conventions:
  - The macros are all related to a main idea 

>   - The file is named for the main idea 

Reviewer 7: Not sure if this is best practice, but seems like if it is a macro used by lots of other macros it should not be buried within another macro file. thoughts?
```suggestion
  - The file is named for the main idea 
  - The macros nested within are not used by other macros
```

Author: Agree - will change this tomorrow!

Author: Actually, I decided to do this since it's the last one. I elaborated a little bit - how is [this](https://github.com/dbt-labs/corp/pull/63/commits/3d5743a86ea8f16147374ef66db4bfbc6037ec8f)?

  |---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
  | Import  | Used to bring data into a model. These are kept relatively simple and refrain from complex operations such as joins and column transformations.                        |
  | Logical | Used to perform a logical step with the data that is brought into the model toward the end result. |

- All `{{ ref() }}` or `{{ source() }}` statements should be placed within import CTEs so that dependent model references are easily seen and located.

- Where applicable, opt for filtering within import CTEs over filtering within logical CTEs. This allows a developer to easily see which data contributes to the end result.

- SQL should end with a simple select statement. All other logic should be contained within CTEs to make stepping through logic easier while troubleshooting.
  Example: `select * from final`

- SQL and CTEs within a model should follow this structure:
  - `with` statement
  - Import CTEs
  - Logical CTEs
  - Simple select statement

### Example SQL with CTEs

>   ``` sql
>   with 
> 
>   -- Import CTEs
>   suppliers as (
>       select * from {{ ref('stg_tpch__suppliers') }}
>   ),
> 
>   nations as (
>       select * from {{ ref('stg_tpch__nations') }}
>   ),
> 
>   regions as (
>       select * from {{ ref('stg_tpch__regions') }}
>   ),
> 
>   -- Logical CTEs
>   locations as (
>       select
>           {{ dbt_utils.surrogate_key([
>               'regions.region_id',            
>               'nations.nation_id'
>           ]) }} as location_sk,
> 
>           regions.region_id,
>           regions.name as region,
>           regions.comment as region_comment,
> 
>           nations.nation_id,
>           nations.name as nation,
>           nations.comment as nation_comment
>         from regions
>       left join nations
>           on regions.region_id = nations.region_id
>   ),
>   
>   final as (
>       select
>           suppliers.supplier_id,
>           suppliers.supplier_name,
>           suppliers.supplier_address,
>           locations.nation,
>           locations.region,
>           suppliers.phone_number,
>           suppliers.account_balance
>       from suppliers
>       inner join locations
>           on suppliers.nation_id = locations.nation_id
>   )
> 
>   -- Simple select statement
>   select * from final
>   ```

Reviewer 7: I am not sure if we have talked about this before, but I am curious about this example CTE SQL. It seems like we are doing a lot of things we dont recommend here — aliasing columns in a marts model, bringing in columns we arent using later. Should we clean this up a bit? 

```suggestion
  ``` sql
  -- Jaffle shop went international!
  with 

  -- Import CTEs
  suppliers as (
      select * from {{ ref('stg_jaffle_shop__suppliers') }}
  ),

  nations as (
      select * from {{ ref('stg_jaffle_shop__nations') }}
  ),

  regions as (
      select * from {{ ref('stg_jaffle_shop__regions') }}
  ),

  -- Logical CTEs
  locations as (
      select
          {{ dbt_utils.surrogate_key([
              'regions.region_id',            
              'nations.nation_id'
          ]) }} as location_sk,

          regions.region_id,
          regions.region,
          regions.region_comment,

          nations.nation_id,
          nations.nation,
          nations.nation_comment
        from regions
      left join nations
          on regions.region_id = nations.region_id
  ),
  
  final as (
      select
          suppliers.supplier_id,
          suppliers.location_id,
          locations.region_id,
          locations.nation_id,
          
          suppliers.supplier_name,
          suppliers.supplier_address,
          suppliers.phone_number,
          locations.region,
          locations.region_comment,
          locations.nation,
          locations.nation_comment,
          
          suppliers.account_balance

      from suppliers
      inner join locations
          on suppliers.location_id = locations.location_sk
  )

  -- Simple select statement
  select * from final
  ```

```

Author: Thanks, this is all really great! I copied this from the template repo we have because I felt like we needed a little more than the example we had before, but could only find this as the shortest example of something that had more than one import & more than one logic step.

I appreciate the attention to detail 🙏 we absolutely should change those things - will get this in on Monday!

Author: Fixed [here](https://github.com/dbt-labs/corp/pull/63/commits/7151ce9c2c40374544f682f398e6ec82582c0076) with a minor change - I reorganized the import CTEs so it lists the data sets in the order they're used

          -- use line breaks to visually separate calculations into blocks
          case
              when my_data.cancellation_date is null
                  and my_data.expiration_date is not null
                  then expiration_data
              when my_data.cancellation_date is null
                  then my_data.start_date + 7
              else my_data.cancellation_date
          end as cancellation_date,

          some_cte_agg.total_field_4,
          some_cte_agg.max_field_5
      from my_data
      left join some_cte_agg  
          on my_data.id = some_cte_agg.id
      where 
          my_data.field_1 = 'abc'
          and (
              my_data.field_2 = 'def'
              or my_data.field_2 = 'ghi'
          )
      qualify row_number() over(
          partition by my_data.field_1
          order by my_data.start_date desc
      ) = 1
  )

  select * from final
  ```

## YAML and Markdown style guide

- Every subdirectory contains their own `.yml` file(s) which contain configurations for the models within the subdirectory.

- YAML and markdown files should be prefixed with an underscore ( `_` ) to keep it at the top of the subdirectory.

> - YAML and markdown files should be named with the convention `_<description>_<config>`.  

Reviewer 7: ```suggestion
- YAML and markdown files should be named with the convention `_<description>__<config>`.  

```

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/43da36c0ea5a7138f6f454a72d1d01fa62be479b)!

>   Examples: `_jaffle_shop_sources.yml`, `_jaffle_shop_docs.md`  

Reviewer 7: ```suggestion
  Examples: `_jaffle_shop__sources.yml`, `_jaffle_shop__docs.md`  

```

Author: Good catch! Will get these in on Monday!

Author: [Done](https://github.com/dbt-labs/corp/pull/63/commits/bcb93bf8b7401ff3cc7de9682bcbe20cb0bf6642). Thank you!

  - `description` is typically the folder of models you're setting configurations for.  
    Examples: `core`, `staging`, `intermediate`
  - `config` is the top-level resource you are configuring.  
    Examples: `docs`, `models`, `sources`
- Indents should use two spaces.

- List items should be indented.

- Use a new line to separate list items that are dictionaries, where appropriate.

- Lines of YAML should be no longer than 80 characters.

- Items listed in a single .yml or .md file should be sorted alphabetically for ease of finding in larger files.

- Each top-level configuration should use a separate `.yml` file (i.e, sources, models)
  Example:
  ```bash
  models
  ├── marts
  └── staging
      └── jaffle_shop

>           ├── _jaffle_shop_docs.md

Reviewer 7: ```suggestion
          ├── _jaffle_shop__docs.md

```

          ├── _jaffle_shop__models.yml
          ├── _jaffle_shop__sources.yml
          ├── stg_jaffle_shop__customers.sql
          ├── stg_jaffle_shop__orders.sql
          └── stg_jaffle_shop__payments.sql
  ```

### Example YAML

>   `_tpch_models.yml`:

Reviewer 7: ```suggestion
  `_jaffle_shop__models.yml`:

```

Author: [This](https://github.com/dbt-labs/corp/pull/63/commits/6d27b7798fb25ba9edc26231a934d268bbce2eb7) and the [previous](https://github.com/dbt-labs/corp/pull/63/commits/8853a4beea585c62c02959be19de1f4fbb641d8c) Done

  ```yaml
  version: 2

  models:
  

>     - name: base_tpch__nations

Reviewer 7: ```suggestion
    - name: base_jaffle_shop__nations

```

Author: Done - gonna refer you back to the commit history if you need to see since there's a lot of little ones here!

      description: This model cleans the raw nations data
      columns:
        - name: nation_id
          tests:
            - unique
            - not_null   

>     - name: base_tpch__regions

Reviewer 7: ```suggestion
    - name: base_jaffle_shop__regions

```

Author: Done

      description: >
        This model cleans the raw regions data before being joined with nations
        data to create one cleaned locations table for use in marts.
      columns:
        - name: region_id
          tests:
            - unique
            - not_null

>     - name: stg_tpch__locations

Reviewer 7: ```suggestion
    - name: stg_jaffle_shop__locations

```

Author: Done

>       description: "{{ doc('tpch_location_details') }}"

Reviewer 7: ```suggestion
      description: "{{ doc('jaffle_shop_location_details') }}"

```

Author: Done

      columns:
        - name: location_sk
          tests:
            - unique
            - not_null
  ```

  ### Example Markdown

>   `_jaffle_shop_docs.md`:

Reviewer 7: ```suggestion
  `_jaffle_shop__docs.md`:

```

Author: Done

| Staging  | Contains models which clean and standardize data        |
| Marts    | Contains models which combine or heavily transform data |
  
Things to note:
- There are different types of models
that typically exist in each of the above categories.  
See [dbt Conventions](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md#dbt-conventions) 
for more information. 

- Read [How we structure our dbt projects](https://docs.getdbt.com/guides/best-practices/how-we-structure/1-guide-overview) to see how we typically structure our projects and further thoughts around organization.

## Modeling Conventions
- Only models in `staging` should select from [sources](https://docs.getdbt.com/docs/building-a-dbt-project/using-sources)
- Models not within the `staging` folder should select from [refs](https://docs.getdbt.com/reference/dbt-jinja-functions/ref).
- The following are the DAG stages that we tend to utilize:
  <details>

  <summary>Common</summary>

    | dag_stage | Typically found in | description                                                        |
    |-----------|--------------------|--------------------------------------------------------------------|

>     | seed_     | /data              | <li> Indicates a data set created from `dbt seed`. |

Reviewer 8: Debated commenting since this is minor, but since the newer projects only name this folder '/seeds', I think we should reference that here to not confuse new users. Either change '/data' to '/seeds' or say '/data or /seeds'

Author: Oof really great catch and *definitely* something we should change!

Author: [Done!](https://github.com/dbt-labs/corp/pull/63/commits/03d6571053c7fda6d73b0dc0cbcb733dd0faa863)

- Aggregations should be executed as early as possible before joining to another table.

- Ordering and grouping by a number (eg. group by 1, 2) is preferred over listing the column names (see [this rant](https://blog.getdbt.com/write-better-sql-a-defense-of-group-by-1/) for why). Note that if you are grouping by more than a few columns, it may be worth revisiting your model design. If you really need to, the [dbt_utils.group_by](https://github.com/dbt-labs/dbt-utils/tree/0.8.6/macros/sql/groupby.sql) function may come in handy.

- Prefer `union all` to `union` [*](http://docs.aws.amazon.com/redshift/latest/dg/c_example_unionall_query.html)

- Avoid table aliases in join conditions (especially initialisms) – it's harder to understand what the table called "c" is compared to "customers".

- If joining two or more tables, _always_ prefix your column names with the table alias. If only selecting from one table, prefixes are not needed.

- Be explicit about your join (i.e. write `inner join` instead of `join`). `left joins` are normally the most common, `right joins` often indicate that you should change which table you select `from` and which one you `join` to.

> Example:

Reviewer 8: Is this Example in the wrong spot? I'd think it'd make more sense to put under line 237 'When dealing with long `when` or `where` clauses, predicates should be on a new line and indented.

Author: Not sure why it shows it in the spot it does here - I pushed a change earlier that shows it [here](https://github.com/dbt-labs/corp/blob/update/proserv_style_guide_changes/dbt_style_guide.md). It's possible you got this one in before I changed it! Either way, it's fixed. Good catch! =)

Reviewer 8: Ha, most likely!

    | report_   | /models/reports    | Indicates that a final data sets are being modeled to pre-aggregate reports for use in outside tooling.                                                                                                                    |

  </details>

## Model Naming

- All objects should be plural.  
  Example: `stg_stripe__invoices.sql` vs. `stg_stripe__invoice.sql`

- All objects should have a prefix to indicate their DAG stage in the flow.  
  See [dbt Conventions](https://github.com/dbt-labs/corp/blob/main/dbt_style_guide.md#dbt-conventions) for more information.

- All models should use the naming convention `<type/dag_stage>_<source/topic>__<additional_context>`. See [this article](https://docs.getdbt.com/blog/stakeholder-friendly-model-names) for more information.
  - For models in the **marts** folder `__<additional_context>` is optional. 
  - Models in the **staging** folder should use the source's name as the `<source/topic>` and the entity name as the `additional_context`.

    Examples:
    - seed_snowflake_spend.csv
    - base_stripe__invoices.sql

>     - stg_stripe__customers.sql

Reviewer 9: It might be helpful to add the example stg_salesforce__customers.sql to demonstrate the need for including the source in the name of every stg model - even when it gets long.  
```suggestion
    - stg_stripe__customers.sql
    - stg_salesforce__customers.sql
```

  ),
  
  suppliers as (
      select * from {{ ref('stg_jaffle_shop__suppliers') }}
  ),
  
  -- Logical CTEs
  locations as (
      select
          {{ dbt_utils.surrogate_key([
              'regions.region_id',            
              'nations.nation_id'
          ]) }} as location_sk,
          regions.region_id,
          regions.region,
          regions.region_comment,
          nations.nation_id,
          nations.nation,
          nations.nation_comment
      from regions
      left join nations
          on regions.region_id = nations.region_id
  ),
  
  final as (
      select
          suppliers.supplier_id,
          suppliers.location_id,
          locations.region_id,
          locations.nation_id,
          suppliers.supplier_name,
          suppliers.supplier_address,
          suppliers.phone_number,
          locations.region,
          locations.region_comment,
          locations.nation,
          locations.nation_comment,
          suppliers.account_balance
      from suppliers
      inner join locations
          on suppliers.location_id = locations.location_sk
  )
  
  -- Simple select statement
  select * from final
  ```

## SQL style guide

> - **DO NOT OPTIMIZE FOR A SMALLER NUMBER OF LINES OF CODE.**  

Reviewer 9: ```suggestion
- **DO NOT OPTIMIZE FOR FEWER LINES OF CODE.**  

```

  - Models in the **staging** folder should use the source's name as the `<source/topic>` and the entity name as the `additional_context`.

    Examples:
    - seed_snowflake_spend.csv
    - base_stripe__invoices.sql
    - stg_stripe__customers.sql
    - stg_salesforce__customers.sql
    - int_customers__unioned.sql
    - fct_orders.sql

## Naming Conventions

- Schema, table and column names should be in `snake_case`.

- Limit use of abbreviations that are related to domain knowledge. An onboarding
  employee will understand `current_order_status` better than `current_os`.

- Use names based on the _business_ terminology, rather than the source terminology.

- Each model should have a primary key that can identify the unique row.

- The primary key of a model should be named `<object>_id`, e.g. `account_id` – this makes it easier to know what `id` is being referenced in downstream joined models.

- If a surrogate key is created, it should be named `<object>_sk`.

- For `base` or `staging` models, columns should be ordered in categories, where identifiers are first and date/time fields are at the end.  
  Example:
  ```sql
  transformed as (

>       select

Reviewer 10: I'm sure there were debates on the newline aspect but I actually prefer newlines.

They give more space for the eyes.

Author: @ericalouie Thanks for commenting this one! It's expected that we'll rehash some debates now that ProServ has aligned and we've opened our thoughts to everyone else. Battle of the teams, commence! ⚔️ 

I'm gonna keep this open so that peeps can 👍 or 👎 =)

Reviewer 2: I personally don't feel strongly about adding newlines after the CTE name or not.

Author: For the time being I'm leaving this as is!

      select
          {{ dbt_utils.surrogate_key([
              'regions.region_id',            
              'nations.nation_id'
          ]) }} as location_sk,
          regions.region_id,
          regions.region,
          regions.region_comment,
          nations.nation_id,
          nations.nation,
          nations.nation_comment
      from regions
      left join nations
          on regions.region_id = nations.region_id
  ),
  
  final as (
      select
          suppliers.supplier_id,
          suppliers.location_id,
          locations.region_id,
          locations.nation_id,
          suppliers.supplier_name,
          suppliers.supplier_address,
          suppliers.phone_number,
          locations.region,
          locations.region_comment,
          locations.nation,
          locations.nation_comment,
          suppliers.account_balance
      from suppliers
      inner join locations
          on suppliers.location_id = locations.location_sk
  )
  
  -- Simple select statement
  select * from final
  ```

## SQL style guide
- **DO NOT OPTIMIZE FOR FEWER LINES OF CODE.**  

  New lines are cheap, brain time is expensive; new lines should be used within  

>   reason to produce code that is easily read.

Reviewer 2: I don't think we need a newline here between within and reason

- Use trailing commas

- Indents should use four spaces. 

- When dealing with long `when` or `where` clauses, predicates should be on a new
  line and indented.  
  Example:
  ```sql
  where 
      user_id is not null
      and status = 'pending'
      and location = 'hq'
  ```

- Lines of SQL should be no longer than 80 characters and new lines should be used to ensure this.  
  Example:
  ```sql

>   sum(
>       case when order_status = 'complete' then order_total end
>   ) as monthly_total,

Reviewer 10: what abt this?

I like having the `case when` inputs as individual lines for readability purposes.
we do the same for window functions so this feeeeels right?

```suggestion
  sum(
      case
          when order_status = 'complete'
              then order_total 
      end
  ) as monthly_total,

```

Author: I think this is a fine ask as it follows our other best practices! I think an hidden rule here is that if it doesn't exceed 80 characters, it's fair game so you're free to be concise, but I think for the purposes of example this should be strict.

      select * from {{ ref('some_cte') }}
  ),

  some_cte_agg as (
      select
          id,
          sum(field_4) as total_field_4,
          max(field_5) as max_field_5
      from some_cte
      group by 1
  ),

  final as (
      select [distinct]
          my_data.field_1,
          my_data.field_2,
          my_data.field_3,

          -- use line breaks to visually separate calculations into blocks
          case
              when my_data.cancellation_date is null
                  and my_data.expiration_date is not null
                  then expiration_data
              when my_data.cancellation_date is null
                  then my_data.start_date + 7
              else my_data.cancellation_date
          end as cancellation_date,

          some_cte_agg.total_field_4,
          some_cte_agg.max_field_5
      from my_data
      left join some_cte_agg  
          on my_data.id = some_cte_agg.id
      where 
          my_data.field_1 = 'abc'
          and (
              my_data.field_2 = 'def'
              or my_data.field_2 = 'ghi'
          )
      qualify row_number() over(
          partition by my_data.field_1
          order by my_data.start_date desc

>       ) = 1

Reviewer 10: ![oh-hullll-yeah](https://user-images.githubusercontent.com/20780760/185723499-9bc8dee5-77d3-407c-bf2b-b20b9dba2fee.JPG)

  - Within the **marts** and **intermediate** folders, `__<additional_context>` is optional. 
  - Models in the **staging** folder should use the source's name as the `<source/topic>` and the entity name as the `additional_context`.

  Examples:
  - base_stripe__invoices.sql
  - stg_stripe__customers.sql
  - int_payments.sql
  - int_customers__unioned.sql
  - fct_orders.sql

### Organization
Our models (typically) fit into three main categories: staging, marts, and base/intermediate.  

For more detail about why we use this structure, check out [this discourse post](https://discourse.getdbt.com/t/how-we-structure-our-dbt-projects/355). The file and naming structures are as follows:
```
├── dbt_project.yml
└── models
    ├── marts
    |   ├── core
    |   |   ├── _docs_core.md
    |   |   ├── _models_core.yml
    |   |   ├── dim_customers.sql
    |   |   └── fct_orders.sql
    |   └── intermediate
    |       ├── _models_int_core.yml
    |       ├── int_payments.sql
    |       ├── int_customers__unioned.sql

>     |       └── int_customers__grouped.sql

Reviewer 5: Everything LGTM! The only comment I have and I don't know how other people feel about this: I am not sure I agree with the base_ and int_ YAML files. I feel like subdirectories are often small enough that a single model_ YAML file can exist for that and its int_ or base_ models. I would recommend it only to very large projects. Interested to hear other peoples' thoughts.

Author: Voting: I'm leaning more towards recommending this vs. not recommending it and keeping it at the parent level. Projects only grow over time, so this is a good setup for future-proofing. However, I really have no strong preference on this. We'll need to change what it says under the yml section further below but keep some examples here as a visual to how these should be named if one decides to use them within further subdirectories.

Reviewer 8: No strong preference on this. But to cast my vote, I like the idea of future-proofing, so leaving them as separate files feels fine to me.

Reviewer 2: I like to have 1 `yml` model file per folder, even if there are only a few models. I find it easier to find where the `yml` file of a specific model is.

Author: Keeping this as is based on comments.

# dbt Style Guide

## Model Naming

- All objects should be plural.  
  Example: `stg_stripe__invoices.sql` vs. `stg_stripe__invoice.sql`

- All objects should have a prefix to indicate their purpose in the flow:
  - `base_` is uncommon, but indicates a process needed before `stg_` - typically when multiple sources are rarely used independently. The stg_ model is then used to join or union the sources together before more robust transformations happen.  
     
     Example: Creating one table for all cleaned location data.
     - `base_location__addresses.sql`, `base_location__countries.sql`, and `base_location__states.sql`  would serve to clean the data and be the 1:1 relationship between the source.
     - `stg_location__locations.sql` would serve to join all location-related data as one entity.
  - `stg_` is used to clean and standardize data before being fundamentally changed in downstream modeling.
  - `int_` is used to indicate a step towards creating a final model surfaced to stakeholders
  - `fct_` is used to indicate a final data set surfaced to stakeholders, and flags data which is in the form of an immutable event stream.

>   - `dim_` is used to indicate a final data set surfaced to stakeholders, and flags data which is used to describe an entity.  

Reviewer 8: Great job on this @christineberger!! I think this makes the `base_` use case much more understandable, as well as all of the use cases for each prefix

├── dbt_project.yml
└── models
    ├── marts
    |   ├── core

>     |   |   ├── _docs.md

Reviewer 11: do we want this to be more specific? `_core_docs.md`? do we have a convention for what to name a markdown file that has shared definitions used across many models?

Reviewer 5: Yeah I originally was fine with _docs.md and then realized I had multiple files open with the same name. I am a fan of _<subdirectory>__docs.md

Reviewer 3: I'm also a fan of context-specific naming, since the Cloud IDE is not always great at delineating which `_docs.yml` file have open. For consistency’s sake, I'd probably offer up `_docs_{subdirectory}.yaml` as an alternative. E.g. `_docs_core.yml`.

Author: 👍 In the IDE when you have a specific file in focus in the editor, the tree highlights where the file is. However, I like this as it’s more apparent without having to know that!

Reviewer: Do we want to do the same with `_models.yml` -> `_models_core.yml`? And `intermediate.yml` -> `_models_core_intermediate.yml`?

    |   |   ├── _models.yml
    |   |   ├── dim_customers.sql
    |   |   └── fct_orders.sql
    |   └── intermediate
    |       ├── intermediate.yml
    |       ├── int_customers__unioned.sql
    |       └── int_customers__grouped.sql
    └── staging
        └── stripe
            ├── base
            |   ├── _models.yml
            |   └── base_stripe__invoices.sql
            ├── _docs.md
            ├── _models.yml

>             ├── _sources.yml

Reviewer 8: 👍  to Lauren's comment

    |       └── int_customers__grouped.sql
    └── staging
        └── stripe
            ├── base
            |   ├── base__stripe_invoices.sql
            ├── src_stripe.yml
            ├── src_stripe.docs
            ├── stg_stripe.yml
            |   ├── _models.yml
            |   └── base_stripe__invoices.sql
            ├── _docs.md
            ├── _models.yml
            ├── _sources.yml
            ├── stg_stripe__customers.sql
            └── stg_stripe__invoices.sql
```
- All objects should be plural, such as: `stg_stripe__invoices`

> - Base tables are prefixed with `base__`, such as: `base__<source>_<object>`

Reviewer 8: I vote to keep the `base` model reference but add more explanation around it. I think the example of base_stripe__customers.sql and base_stripe__deleted_customers.sql is a good one. Maybe the explanation could be something like 'base models are uncommon but can still be used when multiple sources are rarely used independently as a source. The stg_ model is then used to join or union the sources together before more robust transformations happen.'

> - Base tables are prefixed with `base__`, such as: `base__<source>_<object>`

Author: Will put this in the dbt Conventions section!

- Models within the `marts` folder should be materialized as `table` or `incremental`.
  - By default, `marts` should be materialized as `table` within `dbt_project.yml`.
  - If switching to `incremental`, this should be specified in the model's configuration.

## dbt conventions

> - Only models in the `staging` folder should select from [sources](https://docs.getdbt.com/docs/building-a-dbt-project/using-sources)

Author: This states "in the staging folder", which includes base models =) I'll change this to be less confusing!

- Models not in the `staging` folder should select from [refs](https://docs.getdbt.com/reference/dbt-jinja-functions/ref).

## Testing

- At a minimum, unique and not_null tests should be applied to the assumed primary key of each model.

## Naming and field conventions

- Schema, table and column names should be in `snake_case`.

- Use names based on the _business_ terminology, rather than the source terminology.

- Each model should have a primary key that can identify the unique row.

- The primary key of a model should be named `<object>_id`, e.g. `account_id` – this makes it easier to know what `id` is being referenced in downstream joined models.

> - If a surrogate key is created, it should be named `<object>_sk`.

Reviewer 11: totally agree with this but RIP to every table i've ever built with `unique_id` as the primary key 🥲

Reviewer 5: could we also put a note here about limiting abbreviations? Its my personal biggest pet peeve

Reviewer 3: 100% agreed, @SamHarting. If we hold the principle that we should optimize for comprehension over number of lines, then it isn't too big a leap to optimize for comprehension over field name brevity.

Historically, I've used an `<object>_unique_id` naming scheme. Maybe `<object>_surrogate_key`?

Author: 👍

Reviewer 2: I'm all in for avoiding abbreviations in general, but not aligned on replacing `xxx_id` by `xxx_unique_id` or `yyy_sk` by `yyy_surrogate_key`. 

I feel that it jumps to someone's face much more when you look at the last suffix only rather than having to look at multiple words. And if someone doesn't know what an `sk` is, I don't think that calling it `surrogate_key` would help much more.

Author: Added a note on limiting abbreviations! Based on voting in [Slack](https://dbt-labs.slack.com/archives/G01ECQDCG93/p1656093546800559), kept the surrogate key as `<object>_sk`.

> - For `base` or `staging` models, columns should be ordered in categories, where identifiers are first and date/time fields are at the end.  

Reviewer 11: same comment as Benoit above -- we should either remove references to base, to fully explain them and include them

Author: 👍 I can provide an explanation of how best to use each model type

Author: This is notated now in `Model Naming Conventions` under lines 12-17

  Example:
  ```sql
  transformed as (

      select

        -- ids
        order_id,
        customer_id,

>         -- dimensions

Reviewer 11: should this be `varchars` to match numeric/boolean/datetime?

Reviewer 3: > columns should be ordered in categories, where identifiers are first and date/time fields are at the end.

It might be productive to think of this less in terms of data type and more in terms of function. In this example, I see four functions:
1. Identifiers - Primary and foreign keys.
2. Dimensions - Qualitative values used to categorize or segment. A boolean value in this case counts as qualitative.
3. Measures - Quantitative values that you can measure, like the total value of an order
4. Metadata - Data about the data. This would be timestamps or other internal fields.

Using this framework, we get nearly the same order, but with a cleaner abstraction:

```sql
 transformed as (
      select

          -- ids
          order_id,
          customer_id,

          -- dimensions
          order_status,
          is_shipped,

          -- measures
          order_total,
        
          -- metadata
          created_at,
          updated_at,
          _sdc_batched_at

      from source
  )
```
Food for thoughts 🤷

Reviewer 8: 👍  Nicholas's idea seems really straightforward

Reviewer: ^ agree

Author: +1 to @nicholasyager!

Author: Changed this ([commit](https://github.com/dbt-labs/corp/pull/63/commits/47ca66a46843097175638c0c73c845d49059013f)) to regroup by Nicholas' categories.

        order_status,

        -- numeric
        order_total,

        -- boolean
        is_shipped,

        -- date/times
        created_at,
        updated_at,
        _sdc_batched_at

      from source

  )
  ```

- Date/time columns should be named according to these conventions:
  - Timestamps: `<event>_at`
    Format: UTC  
    Example: `created_at`
  
  - Dates: `<event>_date`  
    Format: Date  
    Example: `created_date`

- Booleans should be prefixed with `is_` or `has_`.

- Price/revenue fields should be in decimal currency (e.g. `19.99` for $19.99; many app databases store prices as integers in cents). If non-decimal currency is used, indicate this with suffix, e.g. `price_in_cents`.

> - Avoid reserved words as column names.

Reviewer: Should we add some examples of "reserved words"?

Author: ➕ Changed this [here](https://github.com/dbt-labs/corp/pull/63/commits/11a2e446fc2a50d915dfef78d02120d9d33e6d16)

- Consistency is key! Use the same field names across models where possible.  
Example: a key to the `customers` table should be named `customer_id` rather than `user_id`.

## CTEs

For more information about why we use so many CTEs, check out [this glossary entry](https://docs.getdbt.com/terms/cte).

- Where performance permits, CTEs should perform a single, logical unit of work.

- CTE names should be as verbose as needed to convey what they do.

- CTEs with confusing or noteable logic should be commented with SQL comments, as you would with any complex functions.

- CTEs that are duplicated across models should be pulled out and created as their own models.

- CTEs and SQL within a model should follow this structure:
  - `with` statement
  - Import CTEs
  - Logical CTEs
  - Final CTE
  - Simple `select` statement

- All `{{ ref('...') }}` statements should be placed in CTEs at the top of the file ("import" CTEs)

- SQL should end with a simple select statement. All other logic should be contained within CTEs.  
  Example: `select * from final`

> - Where applicable, opt for filtering within import CTEs over filtering within logical CTEs.

Reviewer 3: Continuing on this thought, there's some research that was conducted last year by Dominik Golebiewski that suggests that in some circumstances it can be considerably more performant to have duplicated import CTEs with separate filters than to have separate logical CTEs that filter from a commont import CTE.

https://medium.com/@AtheonAnalytics/snowflake-query-optimiser-unoptimised-cf0223bdd136

Author: This is where I think we should leave this as our "default style" that we're happy to implement unless we run in to scenarios like the two above. Knowing the things both of you have said is part of our expertise, so I don't feel that we need to elaborate on this because there are the words "where applicable"!

CTE Example:

``` sql
with

events as (

    ...

>     where not _is_deleted

Author: Thank you so much for catching this! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/3c73dfa14a1be5633de3857a88b1cba50d4c7425)

> ## SQL style guide

Reviewer 11: can we change this CTE name to something other than `filtered_events`? confusing name since we're trying to convey that filters should not happen in this CTE

Author: 👍

Author: I missed this one - changed! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/bda4126b5b0b71be6aeb3ab0fa749d008beb8b70)

> - Use trailing commas

Author: +1 to Pat

> - Indents should be two spaces (except for predicates, which should line up with the `where` keyword)

Reviewer 11: 👀 i think my editor defaults to 4 -- thoughts?

Reviewer 5: same... I prefer to be able to just hit tab

Reviewer 3: From a purely pragmatic perspective, there may be a _lot_ a momentum for 4 spaces w/in our codebase already.
But, I'm also biased -- 4 spaces 4 lyfe.

Author: 👍  I’m down for the default, I always found that a little annoying to deal with! In our old version it had 2 and 4 listed in different sections 🤔

Author: Done! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/b5d1dc06dfcac12287e7e844b79ce051ea3d6e00)

- Lines of SQL should be no longer than 80 characters and new lines should be used to ensure this.  
  Example:
  ```sql
  sum(case
    when order_status = 'complete'

>     then order_total

Reviewer 5: personally I think 'then' should be indented. If you have multiple 'when' statements it makes it easier to read

Reviewer 3: For maximum readability:

```sql
sum(
    case
        when order_status = 'complete'
            then order_total
    end
) as monthly_total,
```

Reviewer 2: In that case, I would actually find it clearer to write:
```sql
sum(
    case when order_status = 'complete' then order_total end
) as monthly_total,
```

Reviewer: Maybe better to have a longer case/when statement for this example

Author: Agree - this isn't a hard rule on where to indent, it's talking more about using new lines to make it fit. I do think there's another rule around here that does state it. I'll update this to be a longer case statement.

Author: Changed - [Commit](https://github.com/dbt-labs/corp/pull/63/commits/c48e5e39efd07756943eb8650dce63bfc5c57a3b)

  end) as monthly_total,

  {{ get_windowed_values(
        strategy='sum',
        partition='order_id',
        order_by='created_at',
        column_list=[
          'final_cost'
        ]

>   ) }} as total_final_cost

Reviewer 3: I ❤️  this formatting for macros.

- Aggregations should be executed as early as possible before joining to another table.

- Ordering and grouping by a number (eg. group by 1, 2) is preferred over listing the column names (see [this rant](https://blog.getdbt.com/write-better-sql-a-defense-of-group-by-1/) for why). Note that if you are grouping by more than a few columns, it may be worth revisiting your model design.

- Prefer `union all` to `union` [*](http://docs.aws.amazon.com/redshift/latest/dg/c_example_unionall_query.html)

- Avoid table aliases in join conditions (especially initialisms) – it's harder to understand what the table called "c" is compared to "customers".

- If joining two or more tables, _always_ prefix your column names with the table alias. If only selecting from one table, prefixes are not needed.

- Be explicit about your join (i.e. write `inner join` instead of `join`). `left joins` are normally the most useful, `right joins` often indicate that you should change which table you select `from` and which one you `join` to.

- When filtering by multiple clauses, each clause and expression  should be on a new line.  
Example:
  ```sql
  where user_id is not null

>   and status = 'pending'

Reviewer 11: i would argue for indenting here as well!

Reviewer 5: +

Reviewer 8: Same :)

Reviewer 7: I have always been a fan of new line for each predicate:
```
where 
    user_id is not null
    and status = 'pending'
```

Reviewer 3: Strong agree w/ Pat above.

Author: 👍

Author: There was another comment on this, fixed in the same [commit](https://github.com/dbt-labs/corp/pull/63/commits/c48e5e39efd07756943eb8650dce63bfc5c57a3b). You can read the other feedback [here](https://github.com/dbt-labs/corp/pull/63/files#r905171897).

  and location = 'hq'
  ```

- Joins should list the "left" table first (i.e., the table you're joining data to):
```sql
select
    
    trips.*,
    drivers.rating as driver_rating,
    riders.rating as rider_rating

from trips
left join users as drivers
    on trips.driver_id = drivers.user_id
left join users as riders
    on trips.rider_id = riders.user_id

```

> - *DO NOT OPTIMIZE FOR A SMALLER NUMBER OF LINES OF CODE. NEWLINES ARE CHEAP, BRAIN TIME IS EXPENSIVE*  

Reviewer 3: I would love for this to be in a little "First Principles" section at the top that orients readers to the core philosophy that is guiding this these recommendations. This could easily serve as its core.

Author: I accidently committed a working version of this earlier, but it is now at the top of this section. Let me know if there was something else you had in mind! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/b315a601f908da888a9928e2872815873825fd3b)

my_data as (

    select * from {{ ref('my_data') }}

>     where not _is_deleted

Author: This has been resolved!

my_data as (

    select * from {{ ref('my_data') }}

>     where not is_deleted

Reviewer 8: I've never been a fan of using `where not [boolean]` or even `where [boolean]`- I find it much more confusing and takes me extra time to think through than just seeing `where is_deleted is false` or `where is_deleted is true`. Yes it's a few extra letters to type, but truly infinitely more clear to the reader. 

If it's a performance thing I'm not aware of, then fine, I'm not going to die on this point, but for a readability document I'd vote to change.

Author: I'm okay with changing this, 👍 from me. I'll wait for a few more to chime in here before making the changes!

Author: After looking at this a bit more, I'm going to leave this as is. We're not telling anyone to use the logic in this way, they are free to write `where is_deleted is false` or `where is_deleted = false`.

final as (

    select [distinct]

> 

Reviewer 5: This appears to be a change. I used to do this and then had to train myself out of it. I actually kind of like having the next row be right below select but I can understand why others might like it this way.

Reviewer 8: Agree with Sam. In most cases I don't think extra spacing in the select statement makes it any more readable unless it's a large calculation or case statement. And in most of those cases as long as you utilize multiple lines and indent it's still very readable.

Author: No strong preference on this one from me!

Author: Same - the reason I changed it was that we are add white space at the bottom of the select statement (before the from) and also in between our CTE `()` and the selects within, so changing this made it consistent with the other styling we do. Either way is not a big deal to me, but I think the consistency should show in all the places we do this. For example:

```sql
with

some_item as (
    select * from {{ ref('some_thing') }}
),

other_item as (
    select * from {{ ref('some_other_thing') }}
),

joined as (
    select
        some_item.some_column,
        other_item.some_other_column
    from some_item
    left join other_item
        on some_item.item_id = other_item.item_id
)

select * from joined
```

vs

```sql
with

some_item as (

    select * from {{ ref('some_thing') }}

),

other_item as (

    select * from {{ ref('some_other_thing') }}

),

joined as (

    select

        some_item.some_column,
        other_item.some_other_column

    from some_item
    left join other_item
        on some_item.item_id = other_item.item_id

)

select * from joined
```

Reviewer 8: I personally vote #1 of Christine's comment example. But will be ok with the extra whitespace!

Author: As I said in my previous comment, I have no strong preference on this but for voting's sake I'll choose #1 because:
1. This is how it's laid out throughout the guide and I don't feel like changing it. 😆 
2. In my experience the whitespace does add a breath of fresh air when you get into really big models with lengthy code.

Reviewer 6: i'll vote for #2!

        my_data.field_1,
        my_data.field_2,
        my_data.field_3,

        -- use line breaks to visually separate calculations into blocks
        case

>             

Reviewer: I would argue for indenting here as well!

>             

Author: Changed, let me know if this works! This is in conjunction with some other comments on indentation. 
[Commit](https://github.com/dbt-labs/corp/pull/63/commits/a00e59e2671604618a37bc9b507bffe41c5f3080).

Other commit that fixed indentation:
[Commit](https://github.com/dbt-labs/corp/pull/63/commits/c48e5e39efd07756943eb8650dce63bfc5c57a3b)

    left join some_cte_agg  
        on my_data.id = some_cte_agg.id
    where my_data.field_1 = 'abc'
    and (
        my_data.field_2 = 'def'

>         or my_data.field_2 = 'ghi'

Reviewer 3: ❤️

```

## YAML style guide

- Every subdirectory contains their own `.yml` file(s) which contain configurations for the models within the subdirectory.

- A separate `.yml` file is created per top-level configuration (`sources`, `models`) that applies to the models within the subdirectory. 

- YAML files should be prefixed with an underscore to keep it at the top of the subdirectory.  

>   Example: `_sources.yml`, `_models.yml`

Reviewer: Should we update these to the naming conventions discussed above? _sources_stripe.yml, _models_staging__stripe.yml

Author: Done! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/a2a5e6dde8dd44258c29d38947fe847c7b93a2b6)

> ## Jinja style guide

Reviewer 2: Should we mention that `{%-` and `-%}` can be used to [control whitespace](https://docs.getdbt.com/guides/getting-started/learning-more/using-jinja#use-whitespace-control-to-tidy-up-compiled-code)?

Author: Voting 👍 on this

I would love it if it said something like “Use [Jinja’s whitespace control](https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control) so that rendered SQL can be easily read.” 

I almost feel this is too prescriptive without saying “don’t spend too much time on this”, haha.

This is the part I’m struggling with - 
I want something that’s not _too_ prescriptive but also I feel any good macro is using _some_ whitespace control. 😅

Reviewer 2: I feel like "use whitespace control" with the caveat on not trying to make it perfect is the right balance.

Author: Done - [Commit](https://github.com/dbt-labs/corp/pull/63/commits/c1c88f55fa6a80ca8103325d1de4654ae3941258)

As far as how to whitespace control for for loops, etc, I think we should be careful and ensure we're not making this into a "how to" guide. There are multiple ways you could control the whitespace when making a `for` loop and everyone will do it different (for example, I tend to put {%- if not loop.last %},{% endif %} and others put newlines on the block and add extra whitespace controls). Going to leave this suggestion out.

- Jinja delimiters should have spaces inside of the delimiter between the brackets and your code.  
  Example: `{{ this }}` instead of `{{this}}`
- Use new lines to visually indicate logical blocks of Jinja or to enhance readability.  
  Example:  
  ```jinja 
  {%- set orig_cols = adapter.get_columns_in_relation(ref('fct_orders')) %}

  {%- set new_cols = dbt_utils.star(
        from=ref('fct_order_items'),
        except=orig_cols
  ) %}

  -- original columns
  {%- for col in orig_cols %}

>   {{ col }}

Author: Can you give an example? I'm not really sure what this is stating. =)

>   {{ col }}

Author: I _think_ this is part of a resolved thread - look at [this commit](https://github.com/dbt-labs/corp/pull/63/commits/313f683a97d9e513fa379bb43a89aa82e9b1cbdc) and let me know if this covers it!

>   -- original columns
>   {%- for col in orig_cols %}
>   {{ col }}
>   {% endfor %}

Reviewer 7: suggestion: inside for loops or if blocks, code should be indented
```suggestion
  -- original columns
  {%- for col in orig_cols %}
      {{ col }}
  {% endfor %}

```

Author: 👍

Reviewer 2: I feel like it makes sense in some cases, but in others, with lots of nested `for`/`if`, the compiled code then looks all over the place. If it makes my compiled code easier to read, I would go for 
```jinja
  {%- for col in orig_cols %}
  {{ col }}
  {% endfor %}
```
instead of the indented version.

Reviewer: Last line on this style guide is 

>  Opt for code readability over compiled SQL readability.

So, I would disagree - the indenting makes the jinja code more readable

Author: Changing this to be indented for readability, but agree completely with Benoit. The rendered version gets really weird when you start indenting things all over the place. With that said though, we're going to leave it up to the reader on whether or not they want to indent this and try to adhere to what we're saying in the style guide.

Author: Let me know how this looks everyone! [Commit](https://github.com/dbt-labs/corp/pull/63/commits/313f683a97d9e513fa379bb43a89aa82e9b1cbdc)
